### ISSUE TITLE
Questions about preprocessing custom dataset

### ISSUE THREAD

**Issue by SenYan1999** [andi611/Self-Supervised-Speech-Pretraining-and-Representation-Learning#28]
I am a newcomer to the audio field. I have some questions when use this project to generate the audio embedding for my multimodality model (text and audio)

I want to use Mockingjay, and run `python preprocess_any.py --feature_type=mel' but get 80 dim features, I just simply change num_mel in utility/audio.py from 80 to 160(I see this model need 160dim mel features in README), is it right?

Thanks a lot!

**Comment by andi611** [comment]
Hi,

You don't have to change anything in utility/audio.py.
For the pre-trained Mockingjay model, you need the following arguments:
`python preprocess_any.py --feature_type=mel --delta=True --delta_delta=False --apply_cmvn=True`
The other 80-dim is from delta, mel 80 + delta 80 = the final 160.

**Comment by SenYan1999** [comment]
> Hi,
> 
> You don't have to change anything in utility/audio.py.
> For the pre-trained Mockingjay model, you need the following arguments:
> `python preprocess_any.py --feature_type=mel --delta=True --delta_delta=False --apply_cmvn=True`
> The other 80-dim is from delta, mel 80 + delta 80 = the final 160.

Thank you very much! Another question is when I load Mockingjay model, I just type `transformer = TRANSFORMER(options=options, inp_dim=160)` that changes inp_dim to 160, is it right?

**Comment by andi611** [comment]
You also have to change the content of the `options` dictionary depending on your need:
```python
options = {
    'ckpt_file'     : './result/result_transformer/your_path_to_ckpt/states-1000000.ckpt',
    'load_pretrain' : 'True',
    'no_grad'       : 'True',
    'dropout'       : 'default',
    'spec_aug'      : 'False',
    'spec_aug_prev' : 'True',
    'weighted_sum'  : 'False',
    'select_layer'  : -1,
}
```

**Comment by SenYan1999** [comment]
> Yes, `inp_dim=160`.
> You also have to change the content of the `options` dictionary depending on your need:
> 
> ```python
> options = {
>     'ckpt_file'     : './result/result_transformer/your_path_to_ckpt/states-1000000.ckpt',
>     'load_pretrain' : 'True',
>     'no_grad'       : 'True',
>     'dropout'       : 'default',
>     'spec_aug'      : 'False',
>     'spec_aug_prev' : 'True',
>     'weighted_sum'  : 'False',
>     'select_layer'  : -1,
> }
> ```

OK I see, thank you very much!

**Comment by SenYan1999** [comment]
another question is that how to preprocess the audio data.

1. the audio's time step is different, and i have seen what you have written in TRANSFORMER.process_input_data(), but should i pre-pad 0 to the time step to make the audios in a batch have the same time step? Because if the T is different, I can't build the tensor. But if I do this, how to give the mask information to the model?

2. I have found that my audio is very long. Most of their time step is about 80K or so.
```python
        # forward the whole sequence at once
        if self.max_input_length == 0 or input_len <= self.max_input_length:
            spec_stacked, pos_enc, attn_mask = self.process_input_data(x) # x shape: (B, T, D)
            x = self.model(spec_stacked, pos_enc, attn_mask, output_all_encoded_layers=self.weighted_sum or self.select_layer != -1) # (B, T, D) or # (N, B, T, D)
        # forward the sequence in chunks then concat
        else:
            chunks = torch.chunk(x, chunks=math.ceil(input_len / self.max_input_length), dim=1)
            x_ = []
            for chunk in chunks:
                spec_stacked, pos_enc, attn_mask = self.process_input_data(chunk) # x shape: (B, T, D)
                chunk = self.model(spec_stacked, pos_enc, attn_mask, output_all_encoded_layers=self.weighted_sum or self.select_layer != -1) # (B, T, D) or # (N, B, T, D)
                x_.append(torch.stack(chunk) if type(chunk) is list else chunk)
            x = torch.cat(x_, dim=2 if (self.weighted_sum or self.select_layer != -1) else 1)
```
should i just use the chunk?

**Comment by SenYan1999** [comment]
> another question is that how to preprocess the audio data.
> 
> 1. the audio's time step is different, and i have seen what you have written in TRANSFORMER.process_input_data(), but should i pre-pad 0 to the time step to make the audios in a batch have the same time step? Because if the T is different, I can't build the tensor. But if I do this, how to give the mask information to the model?
> 2. I have found that my audio is very long. Most of their time step is about 80K or so.
> 
> ```python
>         # forward the whole sequence at once
>         if self.max_input_length == 0 or input_len <= self.max_input_length:
>             spec_stacked, pos_enc, attn_mask = self.process_input_data(x) # x shape: (B, T, D)
>             x = self.model(spec_stacked, pos_enc, attn_mask, output_all_encoded_layers=self.weighted_sum or self.select_layer != -1) # (B, T, D) or # (N, B, T, D)
>         # forward the sequence in chunks then concat
>         else:
>             chunks = torch.chunk(x, chunks=math.ceil(input_len / self.max_input_length), dim=1)
>             x_ = []
>             for chunk in chunks:
>                 spec_stacked, pos_enc, attn_mask = self.process_input_data(chunk) # x shape: (B, T, D)
>                 chunk = self.model(spec_stacked, pos_enc, attn_mask, output_all_encoded_layers=self.weighted_sum or self.select_layer != -1) # (B, T, D) or # (N, B, T, D)
>                 x_.append(torch.stack(chunk) if type(chunk) is list else chunk)
>             x = torch.cat(x_, dim=2 if (self.weighted_sum or self.select_layer != -1) else 1)
> ```
> 
> should i just use the chunk?

...after reading nn_transformer.py, i know that just pad 0 to a fixed length is right. Thanks.

**Comment by andi611** [comment]
> another question is that how to preprocess the audio data.
> 
> 1. the audio's time step is different, and i have seen what you have written in TRANSFORMER.process_input_data(), but should i pre-pad 0 to the time step to make the audios in a batch have the same time step? Because if the T is different, I can't build the tensor.

You can first sort the audio according to their length, then put audio with similar length in the same batch. Padding is also required if their length are different, this can be done with `torch.nn.utils.rnn.pad_sequence`.

> But if I do this, how to give the mask information to the model?

At inference time, spectrogram masking is not used.

> 2. I have found that my audio is very long. Most of their time step is about 80K or so.
> should i just use the chunk?

Yes, you can use the chunk, and feed the long input chunk by chunk.

**Comment by SenYan1999** [comment]
> > another question is that how to preprocess the audio data.
> > 
> > 1. the audio's time step is different, and i have seen what you have written in TRANSFORMER.process_input_data(), but should i pre-pad 0 to the time step to make the audios in a batch have the same time step? Because if the T is different, I can't build the tensor.
> 
> You can first sort the audio according to their length, then put audio with similar length in the same batch. Padding is also required if their length are different, this can be done with `torch.nn.utils.rnn.pad_sequence`.
> 
> > But if I do this, how to give the mask information to the model?
> 
> At inference time, spectrogram masking is not used.
> 
> > 1. I have found that my audio is very long. Most of their time step is about 80K or so.
> >    should i just use the chunk?
> 
> Yes, you can use the chunk, and feed the long input chunk by chunk.

It works! Thanks a lot!
